src/apps/lwaftr/fragmentv4.lua

```diff
@@ -0,0 +1,100 @@
+module(..., package.seeall)
+
+local constants = require("apps.lwaftr.constants")
+local packet = require("core.packet")
+local ffi = require("ffi")
+
+local band, bor, lshift = bit.band, bit.bor, bit.lshift
+local ntohs, htons = ffi.C.ntohs, ffi.C.htons
+
+local int16_ptr_t = ffi.typeof("uint16_t*")
+
+local total_length_offset = constants.ethernet_header_size + constants.o_ipv4_total_length
+local frag_id_offset = constants.ethernet_header_size + constants.o_ipv4_identification
+local flags_and_frag_offset_offset = constants.ethernet_header_size + constants.o_ipv4_flags
+local ihl_and_ver_offset = constants.ethernet_header_size + constants.o_ipv4_ver_and_ihl
+
+-- Constants to manipulate the flags next to the frag-offset field directly
+-- as a 16-bit integer, without needing to shift the 3 flag bits.
+local flag_dont_fragment_mask  = 0x4000
+local flag_more_fragments_mask = 0x2000
+local frag_offset_field_mask   = 0x1FFF
+
+
+local function get_u16_at_offset(pkt, offset)
+	return ntohs(ffi.cast(int16_ptr_t, pkt.data + offset)[0])
+end
+
+local function set_u16_at_offset(pkt, offset, value)
+	ffi.cast(int16_ptr_t, pkt.data + offset)[0] = htons(value)
+end
+
+-- TODO: Consider security/performance tradeoffs of randomization
+local internal_frag_id = 0x4242
+local function fresh_frag_id()
+	internal_frag_id = band(internal_frag_id + 1, 0xFFFF)
+	return internal_frag_id
+end
```

Comment:
`frag_id` is a Lua number (64-bit float). I understand you `band` to fetch only the lower 16-bit despite the value might have overflow 16-bit. If I'm correct, val % 0xFFFF would have the same effect, right?

OTOH, I would put `internal_frag_id` inside a closure so it's only accessible from fresh_frag_id:

```lua
local fresh_frag_id = function()
   local internal_frag_id = 0x4242
   return function()
      internal_frag_id = (internal_frag_id + 1) % 0xFFFF
      return internal_frag_id
   end
end